### PR TITLE
734 Additional Performance Tests

### PR DESCRIPTION
**Pull request by sgnn7**
#### What does this PR do (include background context, if relevant)?

This PR adds the following perf agent (juxtaposer) capabilities:
- Ability to test reopened connections
- Ability to test backends with user-set number of parallel threads

Additional features added/changed:
- Added ability to stream-process results asynchronously
- Added ability to stream-process test runs
- Cleaned up timing-related logic
- Made round-based limit be the minimum rounds for all backends (ensures proper comparison)

#### What ticket does this PR close?
Connected to #739 

#### Where should the reviewer start?
[Jenkins Build](https://jenkins.conjur.net/job/cyberark--secretless-broker/job/734-additional-performance-tests/)

#### What is the status of the manual tests?
N/A - this _is_ the testing tool

### CODE REVIEW

File: bin/juxtaposer/main.go [hunk 1]
```diff
@@ -3,167 +3,74 @@ package main
...
-	log.Println("Backends:", len(config.Backends))
 	for backendName, backendConfig := range config.Backends {
 		backendNames = append(backendNames, backendName)
 
 		log.Printf("Setting up backend: %s", backendName)
 
-		if config.Comparison.Type != "sql" {
-			return nil, nil, fmt.Errorf("ERROR: Comparison type supported: %s", config.Comparison.Type)
+		// Sanity check
+		if !strings.HasPrefix(config.Comparison.Type, "sql/") {
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply bin/juxtaposer/main.go]: yeah I guess that could be collapsed 👍

File: bin/juxtaposer/config/config.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+type Comparison struct {
+	BaselineBackend             string `yaml:"baselineBackend"`
+	BaselineMaxThresholdPercent int    `yaml:"baselineMaxThresholdPercent"`
+	Rounds                      string `yaml:"rounds"`
+	Silent                      bool   `yaml:"silent"`
+	Style                       string `yaml:"style"`
+	Threads                     int    `yaml:"threads"`
+	Type                        string `yaml:"type"`
+}
```
Comment (jonahx) [inline bin/juxtaposer/config/config.go]: The names `Type` and `Style` aren't descriptive.

I'd suggest:

- `Type` => `RecreateConnections` (true / false). 
- `Style` => `SqlStatementType`

File: bin/juxtaposer/config/config.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	Password    string `yaml:"password"`
+	Port        string `yaml:"port"`
+	SslMode     string `yaml:"sslmode"`
+	Socket      string `yaml:"socket"`
+	Username    string `yaml:"username"`
+}
+
+type Comparison struct {
+	BaselineBackend             string `yaml:"baselineBackend"`
+	BaselineMaxThresholdPercent int    `yaml:"baselineMaxThresholdPercent"`
```
Comment (jonahx) [inline bin/juxtaposer/config/config.go]: Do we need this?  The stats we were doing yesterday -- % of requests over threshold A, B, C, D, etc -- seems to be a superset of this stat.  

Or would that be too big a change at this point?

File: bin/juxtaposer/config/config.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	err = yaml.Unmarshal(yamlFile, &configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	// Slice out any backends which are ignored
+	filteredBackends := map[string]Backend{}
+	for backendName, backendConfig := range configuration.Backends {
+		if backendConfig.Ignore == false {
```
Comment (jonahx) [inline bin/juxtaposer/config/config.go]: This is probably a change to save for later, but I've given this more thought and I really think cluttering up our config object and the logic which uses it with the `Ignore` field is a mistake.  If users want to ignore something, commenting it out of their config file is not too onerous.  This is something where the convenience of the feature (and I'm not even sure I like the feature)  is nowhere near worth the extra complexity in the code.

File: bin/juxtaposer/formatter/stdout/stdout.go [hunk 5]
```diff
@@ -14,22 +15,22 @@ func NewFormatter(options formatter_api.FormatterOptions) (formatter_api.OutputF
...
-		map[string]string{"name": "Errors", "nameFormat": "%7s", "valueFormat": "%7d"},
+		map[string]string{"name": "Total", "nameFormat": "%18s", "valueFormat": "%18v"},
+		map[string]string{"name": "Rounds", "nameFormat": "%9s", "valueFormat": "%9d"},
+		map[string]string{"name": "Errors", "nameFormat": "%9s", "valueFormat": "%9d"},
 		map[string]string{"name": "Succ %", "nameFormat": "%9s", "valueFormat": "%9.2f"},
 		map[string]string{"name": "Thresh %", "nameFormat": "%9s", "valueFormat": "%9.2f"},
 	}
 
-	dividerString := strings.Repeat("-", 146)
+	dividerString := strings.Repeat("-", 158)
```
Comment (jonahx) [inline bin/juxtaposer/formatter/stdout/stdout.go]: Similar to my comment about rolling your own stats functions, I don't think we should be rolling your own ascii table formatter.  This is stock utility code.  Eg, 

https://github.com/olekukonko/tablewriter

would probably work well.

This change doesn't need to be made now.

File: bin/juxtaposer/config/config.go (hunk 2 above)
Comment (sgnn7) [reply bin/juxtaposer/config/config.go]: Fixed

File: bin/juxtaposer/config/config.go (hunk 3 above)
Comment (sgnn7) [reply bin/juxtaposer/config/config.go]: Yeah without this field, none of the data works anymore ☹️ . We'll need to update the tool to perform the same operations that our python/ruby tools do soon to fix this.

File: bin/juxtaposer/config/config.go (hunk 4 above)
Comment (sgnn7) [reply bin/juxtaposer/config/config.go]: That's fair though to be fair it's only 3 lines. Either way, we can change this later.

File: bin/juxtaposer/formatter/stdout/stdout.go (hunk 5 above)
Comment (sgnn7) [reply bin/juxtaposer/formatter/stdout/stdout.go]: Yup, we can figure out how to do the numbers after we also include the rolling avg for the baseline too.

Comment (jonahx) [reply bin/juxtaposer/formatter/stdout/stdout.go]: I have some architectural ideas around the formatter, but those will def wait for the full review.  Just planting a flag.

File: bin/juxtaposer/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sgnn7) [reply bin/juxtaposer/main.go]: Resolved by followup commits

**Pull request merged**